pkg/kvstore/allocator/allocator_test.go

```diff
@@ -111,6 +115,117 @@ func (s *AllocatorSuite) BenchmarkAllocate(c *C) {
 
 }
 
+func (s *AllocatorSuite) TestRunLocksGC(c *C) {
+	allocatorName := randomTestName()
+	maxID := idpool.ID(256 + c.N)
+	// FIXME: Did this previousy use allocatorName := randomTestName() ? so TestAllocatorKey(randomeTestName())
+	backend1, err := NewKVStoreBackend(allocatorName, "a", TestAllocatorKey(""))
+	c.Assert(err, IsNil)
+	c.Assert(err, IsNil)
+	allocator, err := allocator.NewAllocator(TestAllocatorKey(""), backend1, allocator.WithMax(maxID), allocator.WithoutGC())
+	c.Assert(err, IsNil)
+	shortKey := TestAllocatorKey("1;")
+	var (
+		lock1, lock2 kvstore.KVLocker
+		gotLock1     = make(chan struct{})
+		gotLock2     = make(chan struct{})
+	)
+	go func() {
+		var (
+			err error
+		)
+		lock1, err = backend1.Lock(context.Background(), shortKey)
+		c.Assert(err, IsNil)
+		close(gotLock1)
+		var client kvstore.BackendOperations
+		switch s.backend {
+		case "etcd":
+			client, _ = kvstore.NewClient(context.Background(),
+				s.backend,
+				map[string]string{
+					kvstore.EtcdAddrOption: kvstore.EtcdDummyAddress(),
+				},
+				nil,
+			)
+		case "consul":
+			client, _ = kvstore.NewClient(context.Background(),
+				s.backend,
+				map[string]string{
+					kvstore.ConsulAddrOption:   kvstore.ConsulDummyAddress(),
+					kvstore.ConsulOptionConfig: kvstore.ConsulDummyConfigFile(),
+				},
+				nil,
+			)
+		}
+		lock2, err = client.LockPath(context.Background(), allocatorName+"/locks/"+shortKey.GetKey())
+		c.Assert(err, IsNil)
+		close(gotLock2)
+	}()
+	staleLocks := map[string]kvstore.Value{}
+	staleLocks, err = allocator.RunLocksGC(staleLocks)
+	c.Assert(err, IsNil)
+	c.Assert(len(staleLocks), Equals, 0)
+
+	// Wait until lock1 is gotten.
+	c.Assert(testutils.WaitUntil(func() bool {
+		select {
+		case <-gotLock1:
+			return true
+		default:
+			return false
+		}
+	}, 5*time.Second), IsNil)
+
+	// wait until lock2 is hold
+	time.Sleep(time.Second)
+
+	// Check which locks are stale, it should be lock1 and lock2
+	staleLocks, err = allocator.RunLocksGC(staleLocks)
+	c.Assert(err, IsNil)
+	c.Assert(len(staleLocks), Equals, 2)
+
+	// Get the oldest lock being held
+	var (
+		oldestRev     uint64
+		oldestLeaseID int64
+		sessionID     string
+	)
+	for _, v := range staleLocks {
+		if v.ModRevision < oldestRev {
+			oldestRev = v.ModRevision
+			oldestLeaseID = v.LeaseID
+			sessionID = v.SessionID
+		}
+	}
+	staleLocks[allocatorName+"/locks/"+shortKey.GetKey()] = kvstore.Value{
```

Comment:
Is `staleLocks` intended to be re-created here? It looks like one element is either added/replaced, so presumably it should still have at least 2 elements, which are both deleted immediately afterwards. So then I'm not sure what the waiting at the end of the test is for.. just to ensure that the other goroutine stopped running?

Reply:
When a kvstore client tries to get a lock it will create a key with its lease ID and mod revision. If we have 10 clients trying to get a lock all of them will create a key for each lease so you end up having something like `key/foo/leaseID1`, `key/foo/leaseID2`, etc, if all of them try to get a lock of `key/foo`. As soon client1 unlocks its lock, it will remove `key/foo/leaseID1` and all clients will receive an event stating the key was deleted but only the one with the oldest mod revision can get the lock. If client2 stales this lock, cilium-operator needs to GC the oldest lock of `key/foo` so that client3 can stop being deadlocked. If cilium-operator would delete all keys with the prefix `key/foo`, all clients would lose their lock which is not what we want. Let me know if this was clear.